Implement proxy preferences

micaplush opened the pull request:
This commit adds proxy preferences to the settings screen that are
then used in the ReverseGeocoder, MapBox and public-transport-enabler
(on NetworkProviders that support it, which should apply to every
provider).

One can verify that the app does not establish a direct internet
connection using a tool like nethogs (in Termux).

micaplush committed:
Implement proxy preferences

This commit adds proxy preferences to the settings screen that are
then used in the ReverseGeocoder, MapBox and public-transport-enabler
(on NetworkProviders that support it, which should apply to every
provider).

One can verify that the app does not establish a direct internet
connection using a tool like nethogs (in Termux).

T0astBread commented:
I accept the CLA  ✔️

T0astBread commented:
More on the nethogs thing: If you have a rooted phone you can run nethogs as root in Termux to see when an app establishes a network connection. This is how using a debug build without a proxy looks:

![Screenshot of nethogs showing "de.grobox.liberari..." as one of its entries](https://user-images.githubusercontent.com/24869055/98360748-aee6ab00-202a-11eb-8b40-8402afd84e53.png)

and this is with proxy preferences set to use Tor:

![Screenshot of nethogs not showing anything that looks like the Transportr app but with increased bandwidth usage from Tor](https://user-images.githubusercontent.com/24869055/98360963-0422bc80-202b-11eb-89c3-c12d8e235804.png)

grote commented on app/src/main/java/de/grobox/transportr/settings/SettingsManager.kt:
```diff
@@ -92,6 +96,26 @@ class SettingsManager @Inject constructor(private val context: Context) {
+                val port = Integer.parseInt(settings.getString(PROXY_PORT, null)!!)
+                Proxy(type, InetSocketAddress(InetAddress.getByName(host), port))
+            } catch (e: Exception) {
+                e.printStackTrace()
```
```suggestion
                Log.e(TAG, "Error parsing proxy settings", e)
```

grote commented on app/src/main/java/de/grobox/transportr/utils/TransportrUtils.kt:
```diff
@@ -114,3 +119,16 @@ object TransportrUtils {
+    manager.transportNetwork.value?.let {
+        if (it.networkProvider is AbstractNetworkProvider)
+            (it.networkProvider as AbstractNetworkProvider).setProxy(newProxy)
+    }
```
Should we maybe add the proxy to PTE where its provider gets created instead?

T0astBread replied:
I've tried to do that at first but it seems like the providers are created in a file called `de.grobox.transportr.networks.TransportNetworks` as part of a private val called `networks`, so I'm not sure where to take the proxy configuration from in that case. (Unless I've got something mixed up?)

It would work though (I've done it with hard-coded values) and I agree, it would probably be cleaner.

grote replied:
Maybe add the proxy settings as a lamda parameter here:

https://github.com/grote/Transportr/blob/7b1866e15ae3cb26a31bb69195bc80dcf51c30c1/app/src/main/java/de/grobox/transportr/networks/TransportNetwork.kt#L42

T0astBread replied:
I've tried to implement something like this now but I'm having some problems with it. To not spam this thread, I've made the commit in a new branch: https://github.com/T0astBread/Transportr/commit/4bb422b207aac951530a372bce3ecaf58fcb1f49.

That commit crashes when I do a location search with the following stacktrace:
```
2020-11-17 13:11:12.395 9655-9655/de.grobox.liberario.debug E/AndroidRuntime: FATAL EXCEPTION: main
    Process: de.grobox.liberario.debug, PID: 9655
    kotlin.UninitializedPropertyAccessException: lateinit property settingsManager has not been initialized
        at de.grobox.transportr.locations.LocationView.getSettingsManager$app_debug(LocationView.kt:66)
        at de.grobox.transportr.locations.LocationView$startSuggestLocationsTaskDelayed$1.run(LocationView.kt:239)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6718)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:491)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
```

I think there are a few places in the code where dependency injection won't run before a SettingsManager is needed for the proxy and I'm not really sure what to do about that.

T0astBread replied:
Oh and I've also not made the proxy a parameter to that factory function directly because you'd have to copy+paste `setProxy` in the `TransportNetworks` file for every provider in that case.

grote replied:
Injecting stuff in Views might be trickier than in Activities and Fragments. Thanks for trying. If the first solution also handles network changes, I'd be fine with me, but let's also see what @ialokim thinks.

T0astBread replied:
I've just tried, it seems to work across network changes

ialokim replied:
I think that would be fine with me too, as a network change triggers the MainActivity to be recreated also the proxy settings are applied.

grote reviewed (changes requested):
Awesome! Thanks a lot for your contribution! :smiley: 

Looks like this would close https://github.com/grote/Transportr/issues/265

micaplush committed:
Use Log.e for parsing errors in proxy settings

Co-authored-by: Torsten Grote <grote@users.noreply.github.com>

micaplush committed:
Add myself to contributors

I hope this makes the clabot shut up

micaplush committed:
Fix broken call to Log in SettingsManager

ialokim commented on app/src/main/java/de/grobox/transportr/settings/SettingsManager.kt:
```diff
@@ -92,6 +97,26 @@ class SettingsManager @Inject constructor(private val context: Context) {
+                }
+                val host = settings.getString(PROXY_HOST, null)
+                val port = Integer.parseInt(settings.getString(PROXY_PORT, null)!!)
+                Proxy(type, InetSocketAddress(InetAddress.getByName(host), port))
```
What kind of hosts are accepted here? E.g. do they need `http(s)://`, IP adresses and/or domains?

T0astBread replied:
IP addresses and domains, the protocol and port are handled by other inputs. The default is 127.0.0.1.

ialokim reviewed (changes requested):
Thanks a lot for this PR! I think it would be nice to have some sort of user-feedback right away to see if the proxy settings they entered do work correctly. Not sure though where to hook this in: After a change to some of the fields or before leaving the settings screen?

Would you mind posting a screenshot of how the settings look like with these additions (in active and inactive proxy modes)?

T0astBread commented:
This is what the proxy settings currently look like:

![proxy-settings-on](https://user-images.githubusercontent.com/24869055/99666940-f5e19100-2a6b-11eb-976e-ddc89efa6007.png)
![proxy-settings-off](https://user-images.githubusercontent.com/24869055/99666952-f843eb00-2a6b-11eb-9bed-8ad1526ce25b.png)

As for validation, I think it would make it more clear what's wrong if something happened right after a bad value is entered. Ideally the OK button would be disabled all together when the state of the text input is not acceptable but I'm not sure that's possible with an EditTextPreference. I'll look into how this works.

ialokim commented:
> This is what the proxy settings currently look like

Looks good to me, but I would just prefer a switch over the checkbox for enabling the proxy as documented here: https://source.android.com/devices/tech/settings/settings-guidelines#checkbox

> As for validation, (...) I'll look into how this works.

Thanks for investigating!

micaplush committed:
Use a SwitchPreference for toggling network proxy usage

micaplush committed:
Implement validation for proxy preferences

via the newly created ValidatedEditTextPreference

T0astBread commented:
Implemented some validation now. When the user enters an invalid value for either the host or port the "OK" button on the edit dialog is disabled immediately. The implementation is a bit more complicated than I'd like it to be since Android's EditTextPreference doesn't support that, so I kinda just wrote my own replacement for it.

T0astBread commented:
It should also be noted that the validation doesn't match exactly the same cases as the proxy implementation can handle but, as far as I can tell, the valid cases are a superset of the supported cases and only weird edge cases (like abbreviated IPv4 addresses - "127.1" etc.) are not validated.

ialokim commented on app/src/main/java/de/grobox/transportr/settings/SettingsFragment.kt:
```diff
@@ -91,6 +102,34 @@ class SettingsFragment : PreferenceFragmentCompat() {
+             try/catch is out of the question since it might try to resolve
+             domain names.
+             */
+            InternetDomainName.isValid(value) || HostSpecifier.isValid(value)
```
Couldn't we use something like [`Patterns.DOMAIN_NAME`](https://developer.android.com/reference/android/util/Patterns#DOMAIN_NAME) and `Patterns.IP_ADDRESS`? For some reason the current check accepts "g" as host for me which doesn't make too much sense I'd guess

T0astBread replied:
Thanks, I didn't know about those. Although "g" should theoretically be okay since it's a valid hostname and it could theoretically also resolve. This is also how localhost works. (And "localhost" is also not accepted with Patterns.)

ialokim commented on app/src/main/java/de/grobox/transportr/utils/TransportrUtils.kt:
```diff
@@ -114,3 +119,16 @@ object TransportrUtils {
 
 fun Location.hasLocation() = hasCoord() && (latAs1E6 != 0 || lonAs1E6 != 0)
+
+fun updateGlobalHttpProxy(newProxy: Proxy, manager: TransportNetworkManager) {
```
Might be better to move this inside the `TransportrUtils` object (static class).

ialokim reviewed (changes requested):
Thanks for adding the validation part!

I've just tested your changes inserting an invalid hostname (in this case only the letter "g"). The validation doesn't complain about this (strange enough?), but also no other feedback is given that the proxy connection could not be established. I was able to make network requests as before over the normal internet connection without further notice.

I think the best thing would be to somehow test the proxy settings by performing a request after leaving the settings screen and report invalid values to the user straight away. At least no further network requests should be sent through the normal connection without informing the user.

micaplush committed:
Use Patterns for proxy host validation

micaplush committed:
Test proxy connection on proxy preference changes

This also fixes a bug where stale proxy settings were sometimes used
since PreferenceChangeListeners fire before the settingsManager of
the SettingsFragment sees the new values in its settings instance.

T0astBread commented:
Finally another update from me! Proxy connections are tested after the user changes a proxy preference now. If the test fails, an AlertDialog with the error message is shown. Preferences are still accepted on a failed test since it would otherwise be impossible to change e.g. the host and port at the same time.

ialokim commented on app/src/main/java/de/grobox/transportr/settings/SettingsFragment.kt:
```diff
@@ -91,6 +103,37 @@ class SettingsFragment : PreferenceFragmentCompat() {
+                        TransportrUtils.updateGlobalHttpProxy(newProxy, manager)
+                    } catch (e: Exception) {
+                        Log.e(TAG, "Invalid proxy settings: " + e.message)
+                        AlertDialog.Builder(context!!)
```
I would prefer a less disturbing Toast message at this point.

T0astBread replied:
Do you think it would be a good idea to handle some common errors using a Toast with a simplified message and use an AlertDialog with the Exception's message for anything unexpected?

ialokim commented on app/src/main/java/de/grobox/transportr/settings/SettingsFragment.kt:
```diff
@@ -91,6 +103,37 @@ class SettingsFragment : PreferenceFragmentCompat() {
+        }
+
+        findPreference<ValidatedEditTextPreference>(PROXY_HOST)!!.validate = { value ->
+            Patterns.DOMAIN_NAME.matcher(value).matches() || Patterns.IP_ADDRESS.matcher(value).matches()
```
Perhaps we should add "localhost" then manually as I would see it as a common setting?

ialokim commented on app/src/main/java/de/grobox/transportr/settings/SettingsManager.kt:
```diff
@@ -92,6 +100,23 @@ class SettingsManager @Inject constructor(private val context: Context) {
         }
 
+    @Throws(UnknownHostException::class, IllegalStateException::class)
+    fun getProxy(proxyPrefOverrides: Map<String, Any>): Proxy {
```
All uses of this function seem to set at most one new setting. Why not simply using a `Pair<String, Any>` here?

T0astBread replied:
You're right but I think the code is actually more readable with the map compared to a single `Pair<String, Any>?`, if we want to handle mismatched types the same way.

Compare
```kotlin
val isEnabled = proxyPrefOverrides[PROXY_ENABLE] as Boolean? ?: settings.getBoolean(PROXY_ENABLE, false)
```
to
```kotlin
val isEnabled =
    if (proxyPrefOverride?.first == PROXY_ENABLE && proxyPrefOverride.second is Boolean)
        proxyPrefOverride.second as Boolean
    else
        settings.getBoolean(PROXY_ENABLE, false)
```
or (avoiding the explicit cast)
```kotlin
val overrideKey = proxyPrefOverride?.first
val overrideVal = proxyPrefOverride?.second
val isEnabled = if (overrideKey == PROXY_ENABLE && overrideVal is Boolean)
    overrideVal else settings.getBoolean(PROXY_ENABLE, false)
```

ialokim reviewed (changes requested):
Sorry for the delay. I've finally had another look at your changes. Please see my comments in-code.

I still think it would be nice to finally test the proxy settings when leaving the settings screen (`onBackButtonPressed` or something similar) and in case the connection could not be established show another Toast message and deactivate the proxy again.

micaplush committed:
Explicitly allow "localhost" for proxy host preference

T0astBread commented:
I've avoided automatically deactivating the proxy since I figured someone might not notice the Toast and accidentally use the app without a proxy. If the app just continues using invalid proxy settings (which blocks all connections), there's no way a user won't notice something's wrong.

grote commented:
I don't have time at the moment to do a review, so just a comment: When setting a proxy, I'd show a progress bar until we know the proxy works and then only set the setting for real when it does. Otherwise don't set the settings and show an error dialog.

T0astBread commented:
Where should the progress bar be placed? At the top or bottom of the settings activity, perhaps together with a Snackbar displaying something like "Checking proxy settings..."? Or a progress dialog (those are deprecated)? And what should happen when the user exits the settings activity while the proxy connection is being checked?

I could add another set of proxy preferences that act as "untested" proxy preferences, which are what the user actually sees in the UI. Then, when the user exits the settings activity and the untested settings differ from the currently used ones, the untested settings are checked with a non-blocking progress bar. If the user presses back to exit the activity a Toast shows up saying something like "Press back again to cancel the proxy check". Something similar is probably also possible for the back arrow in the app bar.

But is that really necessary? I know, blocking the UI for a network request is really bad UX but who, other than power users, would mess with proxy settings anyways?

grote commented:
The network request would have to be done off the UI thread of course, but the UI could still artificially be blocked. Just show a Dialog fragment with a progress bar and a text saying "testing proxy settings". the same fragment could even show the error message and attach to the same ViewModel as the settings.

ialokim commented:
I've just have done some comparison on how other apps handle proxy settings:
- F-Droid and Aurora Store simply crash after changing to invalid proxy settings, I had to clear the application data to get them to work again - definitely not the way to go.
- DAVx5 at least doesn't crash, but simply reports a general network error as it would do whenever the Cal/CardDav server would not be reachable.
- Telegram FOSS is the only app I've tested that actually gives the user some feedback directly to the proxy settings. See the following screen record:

https://user-images.githubusercontent.com/18088991/108270436-bceeb080-716f-11eb-9632-d99e4e5e0bda.mp4

I think it would make sense to move the proxy settings to a new "sub-preference" screen similar to how Telegram does it. Of course I would argue one proxy setting at a time should be enough. But then we could really check the proxy settings before returning to the main preference screen and prevent "false" or non-working settings. Also, in the main preference screen, there could be a description field stating whether the proxy settings are working and enabled or not.

T0astBread commented:
So, something like this?

![Diagram describing user interface states: A settings screen with a link to a detailed "Proxy Settings" screen. The proxy settings screen checks, upon clicking a "save" check mark, the entered proxy settings and either returns the user to the previous screen or displays an error. The settings screen also indicates if the proxy settings work with a non-blocking connection status label.](https://user-images.githubusercontent.com/24869055/108297093-099bb100-719b-11eb-9eac-fa533b37f400.png)

ialokim commented:
Yes, that looks really good to me! Just two minor things:
- Perhaps we could combine the switch and the area to go to the proxy settings into one single row, as e.g. in Android system's settings for Wifi. See [this](https://stackoverflow.com/questions/50433203/how-to-properly-customize-switchpreferencecompat-so-that-it-behaves-like-wi-fi-s) for an image, but not sure if the answer helps much. I've also found [this](https://gist.github.com/marchold/45e22839eb94aa14dfb5) but I could imagine that something similar might exist natively by now.
- The check should also be triggered when going back from the proxy settings to the main settings using the back arrow. Not sure if a dedicated check mark in the top right would be necessary then.

grote commented:
@T0astBread thanks for the mockup, this looks great!

I personally don't think that we need these colored status information on the main settings screen for proxy settings. Just say if a proxy is set or if it isn't.

If we allow the user to leave the proxy screen while still checking the settings, they either won't see the result (success or fail) or we don't set the  proxy settings at all. The first option is probably better, but then it might be overkill to add checking status info to the main screen. Just update the proxy livedata when we have the result.
